include/trial/protocol/json/partial/copy_and_look_ahead.hpp

```diff
@@ -0,0 +1,176 @@
+#ifndef TRIAL_PROTOCOL_JSON_PARTIAL_COPY_AND_LOOK_AHEAD_HPP
+#define TRIAL_PROTOCOL_JSON_PARTIAL_COPY_AND_LOOK_AHEAD_HPP
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (C) 2021 Vinícius dos Santos Oliveira <vini.ipsmaker@gmail.com>
+//
+// Distributed under the Boost Software License, Version 1.0.
+//    (See accompanying file LICENSE_1_0.txt or copy at
+//          http://www.boost.org/LICENSE_1_0.txt)
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include <trial/protocol/json/partial/skip.hpp>
+
+namespace trial
+{
+namespace protocol
+{
+namespace json
+{
+namespace partial
+{
+
+namespace detail {
+template<class CharT>
+struct CopyAndLookAheadCollector
+{
+    using view_type = typename basic_reader<CharT>::view_type;
+    using size_type = typename basic_reader<CharT>::size_type;
+
+    CopyAndLookAheadCollector(view_type key)
+        : pattern{key}
+    {}
+
+    void push_back(CharT v)
+    {
+        if (!process) return;
+
+        if (!pattern.starts_with(v)) {
+            process = false;
+            return;
+        }
+
+        pattern.remove_prefix(1);
+    }
+
+    void append(const CharT *data, size_type size)
+    {
+        if (!process) return;
+
+        view_type run{data, size};
+        if (!pattern.starts_with(run)) {
+            process = false;
+            return;
+        }
+
+        pattern.remove_prefix(size);
+    }
+
+    bool current_value_matches_pattern()
+    {
+        return process && pattern.size() == 0;
+    }
+
+    bool process = true;
+    view_type pattern;
+};
+} // namespace detail
+
+template<class CharT, class F>
+auto/*=bool*/ copy_and_look_ahead(const basic_reader<CharT> &reader_,
```

Comment:
As we are going the copy the reader anyways, we might as well pass it by value and drop "copy_and_" from the name. This also enables the user to do a "destructive look-ahead" by moving the reader when calling this function.

Reply:
> As we are going the copy the reader anyways, we might as well pass it by value

Done: https://github.com/breese/trial.protocol/pull/57/commits/76c94643aaa2e7243a322ad8703b4319548bd2df

> This also enables the user to do a "destructive look-ahead" by moving the reader when calling this function.

Indeed. It's a good suggestion.

> and drop "copy_and_" from the name.

I'm not sure about this one. So far every `partial::*` algorithm modified its input argument. It'd be good to have a name hint that something funny is going on. Do you have another suggestion for a name hint? I couldn't think of a better hint.

Conceptually we're still copying. It just happens that we might construct our copy by move. So this excuse can be used.

Reply:
You have a good point that we need to clearly separate between algorithms that modifies the reader and those that do not.

We could let the immutable algorithms take a view instead of a reader. The algorithm may decide to create a reader, but that becomes an implementation detail. Something like this:
```
template <typename View>
bool contains(View view);

template <typename View>
View find_value(View view, View key);
```

Reply:
> We could let the immutable algorithms take a view instead of a reader. The algorithm may decide to create a reader, but that becomes an implementation detail.

That still doesn't tickle right to me. The type becomes invisible when we're calling a function.

```c
foobar(x);
```

So the type isn't a real hint to help readability.

However this discussion reminds me of `std::ref()` vs `std::cref()`. We're already using the namespace name to hint expectations, so maybe we could create a new namespace for algorithms such as `look_ahead()`. Maybe `cpartial` (not the best name, but maybe we can think of a better one soon)?